util/offline_db/get_resources_from_db.py

```diff
@@ -0,0 +1,223 @@
+
+    try:
+        response = request.urlopen(req)
+    except Exception as e:
+        raise Exception(f"Failed to get resources: {e}")
+
+    return json.loads(response.read().decode("utf-8"))["documents"]
+
+
+def save_resources_to_file(resources: List, output: pathlib.Path):
+    """
+    This function saves all the JSON objects for resources to a file.
+    :param resources: List of JSON objects for resources
+    :param output: Output directory absolute path
+    """
+
+    path = output.joinpath("resources.json")
+
+    with open(path, "w") as f:
+        json.dump(resources, f, indent=4)
+
+
+def progress_hook(t):
+    last_b = [0]
+
+    def update_to(b=1, bsize=1, tsize=None):
+        if tsize not in (None, -1):
+            t.total = tsize
+        displayed = t.update((b - last_b[0]) * bsize)
+        last_b[0] = b
+        return displayed
+
+    return update_to
+
+
+def download_resources(output: pathlib.Path):
+    """
+    This function downloads the resources which have a url field and
+    updates the url field to point to the local download of the resource.
+    :param output: Output directory absolute path
+    """
+
+    path = output.joinpath("offline_resources")
+    if not path.exists():
+        path.mkdir()
+
+    with open(output.joinpath("resources.json")) as f:
+        resources = json.load(f)
+
+    for resource in resources:
+        if "url" in resource.keys():
+            url = resource["url"]
+            filename = pathlib.Path(
+                *pathlib.Path(parse.urlsplit(url).path).parts[3:]
+            )
+            filepath = pathlib.Path(path).joinpath(filename)
+            filepath.parent.mkdir(parents=True, exist_ok=True)
+
+            if (
+                not filepath.exists()
+                or hashlib.md5(filepath.read_bytes()).hexdigest()
+                != resource["md5sum"]
+            ):
+                with tqdm(
+                    unit="B",
+                    unit_scale=True,
+                    unit_divisor=1024,
+                    miniters=1,
+                    desc=f"Downloading {url}",
+                ) as t:
+                    request.urlretrieve(url, filepath, progress_hook(t))
+            resource["url"] = filepath.absolute().as_uri()
+
+    with open(output.joinpath("resources.json"), "w") as f:
+        json.dump(resources, f, indent=4)
+
+
+if __name__ == "__main__" or __name__ == "__m5_main__":
+    parser = argparse.ArgumentParser(
+        description="Get resources from the database"
+    )
+    parser.add_argument(
+        "--output_dir",
+        type=str,
+        default=pathlib.Path.cwd(),
+        help="Output directory absolute path, default is the cwd."
+        "The resources.json and all resources will be saved in this directory",
+    )
+    args = parser.parse_args()
+
+    output_path = pathlib.Path(args.output_dir)
+
+    # Get the gem5 config from the stable branch of gem5
+    gem5_config = request.urlopen(
```

Comment:
As a user I want this script to download the resources needed from my gem5 repo. This repo might be a few version behind upstream release.
In your current solution, you always get the resources from upstream stable.
A better way IMO would be to have an option:

```
parser.add_argument(
    "--config-file",
   required=True,
   help="Path to the config file used to get list of clients providing resources. Default one is src/python/gem5_default_config.py")
```

Related question: why is this file a .py and not .json?

Reply:
The config file is .py because we wanted to build gem5 with the default config so we can import it during initialization in `client.py` and having a JSON file was causing issues with building with scons so we switched it to a .py.

I am working on the documentation to update config files, the current version can be found here: https://github.com/gem5/website/pull/90